File: src/go/pkg/netipc/service/raw/cgroups_lookup.go
package raw

import (
	"bytes"
	"errors"

	"github.com/netdata/netdata/go/plugins/pkg/netipc/protocol"
)

// CgroupsLookupHandler serves a single CGROUPS_LOOKUP service kind.
type CgroupsLookupHandler func(*protocol.CgroupsLookupRequestView, *protocol.CgroupsLookupBuilder) bool

func cgroupsLookupRequestSize(paths [][]byte) (int, error) {
	return cgroupsLookupRequestSizeForCount(len(paths), func(i int) int {
		return len(paths[i])
	})
}

func cgroupsLookupRequestSizeForLengths(pathLens []int) (int, error) {
	return cgroupsLookupRequestSizeForCount(len(pathLens), func(i int) int {
		return pathLens[i]
	})
}

func cgroupsLookupRequestSizeForCount(count int, pathLenAt func(int) int) (int, error) {
	dirSize, err := checkedLookupMul(count, protocol.LookupDirEntrySize)
	if err != nil {
		return 0, err
	}
	size, err := checkedLookupAdd(protocol.CgroupsLookupReqHdr, dirSize)
	if err != nil {
		return 0, err
	}
	data := size
	for i := range count {
		data, err = checkedLookupAlign8(data)
		if err != nil {
			return 0, err
		}
		data, err = checkedLookupAdd(data, pathLenAt(i))
		if err != nil {
			return 0, err
		}
		data, err = checkedLookupAdd(data, 1)
		if err != nil {
			return 0, err
		}
	}
	return data, nil
}

func cgroupsLookupNextRequest(paths [][]byte, maxPayload uint32) (int, int, error) {
	if len(paths) == 0 {
		size, err := cgroupsLookupRequestSize(nil)
		return 0, size, err
	}
	if maxPayload == 0 {
		maxPayload = protocol.MaxPayloadDefault
	}
	lo, hi := 1, len(paths)
	bestCount := 0
	bestSize := 0
	for lo <= hi {
		mid := lo + (hi-lo)/2
		size, err := cgroupsLookupRequestSize(paths[:mid])
		if err != nil {
			return 0, 0, err
		}
		if lookupSizeWithinLimit(size, maxPayload) {
			bestCount = mid
			bestSize = size
			lo = mid + 1
			continue
		}
		hi = mid - 1
	}
	if bestCount == 0 {
		return 0, 0, protocol.ErrOverflow
	}
	return bestCount, bestSize, nil
}

func cgroupsLookupOversizedRequestItem(path []byte) ([]byte, error) {
	size, err := cgroupsLookupOversizedRequestItemSize(len(path))
	if err != nil {
		return nil, err
	}

	buf := make([]byte, size)
	builder := protocol.NewCgroupsLookupBuilder(buf, 1, 0)
	if err := builder.Add(protocol.CgroupLookupOversizedItem, 0, path, nil, nil); err != nil {
		return nil, err
	}
	n := builder.Finish()
	view, err := protocol.DecodeCgroupsLookupResponse(buf[:n])
	if err != nil {
		return nil, err
	}
	rawItem, err := view.RawItem(0)
	if err != nil {
		return nil, err
	}
	return cloneLookupRawItem(rawItem), nil
}

func cgroupsLookupOversizedRequestItemSize(pathLen int) (int, error) {
	size, err := checkedLookupAdd(protocol.CgroupsLookupRespHdr, protocol.LookupDirEntrySize)
	if err != nil {
		return 0, err
	}
	size, err = checkedLookupAlign8(size)
	if err != nil {
		return 0, err
	}
	itemSize, err := checkedLookupAdd(protocol.CgroupsLookupItemHdr, pathLen)
	if err != nil {
		return 0, err
	}
	itemSize, err = checkedLookupAdd(itemSize, 2)
	if err != nil {
		return 0, err
	}
	return checkedLookupAdd(size, itemSize)
}

// CallCgroupsLookup performs a blocking typed CGROUPS_LOOKUP call.
// The returned view is valid until the next typed call on this client.
func (c *Client) CallCgroupsLookup(paths [][]byte) (*protocol.CgroupsLookupResponseView, error) {
	return c.CallCgroupsLookupWithTimeout(paths, 0)
}

func (c *Client) CallCgroupsLookupWithTimeout(paths [][]byte, timeoutMs uint32) (*protocol.CgroupsLookupResponseView, error) {
	if err := c.validateMethod(protocol.MethodCgroupsLookup); err != nil {
		return nil, err
	}
	if uint64(len(paths)) > uint64(c.maxLogicalLookupItems) {
		return nil, protocol.ErrOverflow
	}
	if err := c.ensureReadyForLogicalLookup(); err != nil {
		return nil, err
	}

	expectedCount, err := checkedLookupU32(len(paths))
	if err != nil {
		return nil, err
	}
	rawItems := make([][]byte, 0, len(paths))
	start := 0
	var generation uint64
	haveGeneration := false
	subcalls := uint32(0)
	deadline := c.newLookupDeadline(timeoutMs)
	for {
		reqCount, reqSize, err := cgroupsLookupNextRequest(paths[start:], c.sessionMaxRequestPayloadBytes())
		if err != nil {
			if err == protocol.ErrOverflow && start < len(paths) {
				oneItemSize, serr := cgroupsLookupRequestSize(paths[start : start+1])
				if serr == nil {
					if c.ensureLookupRequestCapacity(oneItemSize) == nil {
						continue
					}
				}
				rawItem, serr := cgroupsLookupOversizedRequestItem(paths[start])
				if serr != nil {
					return nil, serr
				}
				rawItems = append(rawItems, rawItem)
				start++
				continue
			}
			return nil, err
		}
		reqPaths := paths[start : start+reqCount]
		reqBuf := ensureClientScratch(&c.requestBuf, reqSize)
		reqLen, err := protocol.EncodeCgroupsLookupRequest(reqPaths, reqBuf)
		if err != nil {
			return nil, err
		}

		remainingTimeout, terr := c.lookupRemainingTimeout(deadline)
		if terr != nil {
			return nil, terr
		}
		subcalls++
		if subcalls > c.maxLogicalLookupSubcalls {
			return nil, protocol.ErrOverflow
		}
		var payload []byte
		if err := c.callWithRetry(func() error {
			_, callPayload, rerr := c.doRawCallWithTimeout(protocol.MethodCgroupsLookup, reqBuf[:reqLen], remainingTimeout)
			if rerr != nil {
				return rerr
			}
			payload = callPayload
			return nil
		}); err != nil {
			return nil, err
		}
		view, derr := protocol.DecodeCgroupsLookupResponse(payload)
		if derr != nil {
			return nil, derr
		}
		if !haveGeneration {
			generation = view.Generation
			haveGeneration = true
		} else if view.Generation != generation {
			return nil, protocol.ErrBadLayout
		}
		reqCountU32, err := checkedLookupU32(len(reqPaths))
		if err != nil {
			return nil, err
		}
		if view.ItemCount != reqCountU32 {
			return nil, protocol.ErrBadItemCount
		}
		payloadExceededAt := -1
		for i, expected := range reqPaths {
			itemIndex, ierr := checkedLookupU32(i)
			if ierr != nil {
				return nil, ierr
			}
			item, ierr := view.Item(itemIndex)
			if ierr != nil {
				return nil, ierr
			}
			if !bytes.Equal(item.Path.Bytes(), expected) {
				return nil, protocol.ErrBadLayout
			}
			if item.Status == protocol.CgroupLookupPayloadExceeded {
				payloadExceededAt = i
				break
			}
			rawItem, ierr := view.RawItem(itemIndex)
			if ierr != nil {
				return nil, ierr
			}
			rawItems = append(rawItems, cloneLookupRawItem(rawItem))
		}
		if payloadExceededAt < 0 {
			start += reqCount
			if start < len(paths) {
				continue
			}
			rawItemCount, cerr := checkedLookupU32(len(rawItems))
			if cerr != nil {
				return nil, cerr
			}
			if rawItemCount != expectedCount {
				return nil, protocol.ErrBadItemCount
			}
			size, serr := lookupRawResponseSize(protocol.CgroupsLookupRespHdr, rawItems)
			if serr != nil {
				return nil, serr
			}
			if lookupSizeExceedsLimit(size, c.maxLogicalLookupResponseBytes) {
				return nil, protocol.ErrOverflow
			}
			stitched := make([]byte, size)
			n, eerr := protocol.EncodeCgroupsLookupRawResponse(rawItems, generation, stitched)
			if eerr != nil {
				return nil, eerr
			}
			return protocol.DecodeCgroupsLookupResponse(stitched[:n])
		}
		for i := payloadExceededAt; i < len(reqPaths); i++ {
			itemIndex, ierr := checkedLookupU32(i)
			if ierr != nil {
				return nil, ierr
			}
			item, ierr := view.Item(itemIndex)
			if ierr != nil {
				return nil, ierr
			}
			if !bytes.Equal(item.Path.Bytes(), reqPaths[i]) ||
				item.Status != protocol.CgroupLookupPayloadExceeded {
				return nil, protocol.ErrBadLayout
			}
		}
		if payloadExceededAt == 0 {
			return nil, protocol.ErrOverflow
		}
		start += payloadExceededAt
	}
}

// CgroupsLookupDispatch adapts a typed cgroups lookup handler to the raw dispatch shape.
func CgroupsLookupDispatch(handle CgroupsLookupHandler) DispatchHandler {
	if handle == nil {
		return nil
	}
	return func(request []byte, responseBuf []byte) (int, error) {
		n, err := protocol.DispatchCgroupsLookup(request, responseBuf, func(req *protocol.CgroupsLookupRequestView, builder *protocol.CgroupsLookupBuilder) bool {
			return handle(req, builder)
		})
		if errors.Is(err, protocol.ErrHandlerFailed) {
			return 0, errHandlerFailed
		}
		return n, err
	}
}

File: src/crates/netipc/src/service/raw/cgroups_lookup.rs
use super::client::{ClientConfig, RawCallKind, RawClient};
use super::common::lookup_raw_response_size;
use super::dispatch::{dispatch_error_from_protocol, DispatchHandler};
use crate::protocol::{
    self, CgroupsLookupBuilder, CgroupsLookupRequestView, CgroupsLookupResponseView, NipcError,
    CGROUPS_LOOKUP_ITEM_HDR_SIZE, CGROUPS_LOOKUP_REQ_HDR_SIZE, CGROUPS_LOOKUP_RESP_HDR_SIZE,
    CGROUP_LOOKUP_OVERSIZED_ITEM, CGROUP_LOOKUP_PAYLOAD_EXCEEDED, LOOKUP_DIR_ENTRY_SIZE,
    METHOD_CGROUPS_LOOKUP,
};
use std::sync::Arc;

fn cgroups_lookup_request_size(paths: &[&[u8]]) -> Result<usize, NipcError> {
    let dir_size = paths
        .len()
        .checked_mul(LOOKUP_DIR_ENTRY_SIZE)
        .ok_or(NipcError::Overflow)?;
    let mut data = CGROUPS_LOOKUP_REQ_HDR_SIZE
        .checked_add(dir_size)
        .ok_or(NipcError::Overflow)?;
    for path in paths {
        data = data
            .checked_add(7)
            .map(|v| v & !7)
            .ok_or(NipcError::Overflow)?;
        data = data
            .checked_add(path.len())
            .and_then(|v| v.checked_add(1))
            .ok_or(NipcError::Overflow)?;
    }
    Ok(data)
}

fn cgroups_lookup_next_request_count(
    paths: &[&[u8]],
    max_payload: u32,
) -> Result<(usize, usize), NipcError> {
    if paths.is_empty() {
        return Ok((0, cgroups_lookup_request_size(&[])?));
    }
    let cap = if max_payload == 0 {
        protocol::MAX_PAYLOAD_DEFAULT
    } else {
        max_payload
    } as usize;
    let mut lo = 1usize;
    let mut hi = paths.len();
    let mut best_count = 0usize;
    let mut best_size = 0usize;
    while lo <= hi {
        let mid = lo + (hi - lo) / 2;
        let size = cgroups_lookup_request_size(&paths[..mid])?;
        if size <= cap {
            best_count = mid;
            best_size = size;
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    if best_count == 0 {
        return Err(NipcError::Overflow);
    }
    Ok((best_count, best_size))
}

fn cgroups_lookup_oversized_request_item(path: &[u8]) -> Result<Vec<u8>, NipcError> {
    let mut size = CGROUPS_LOOKUP_RESP_HDR_SIZE
        .checked_add(LOOKUP_DIR_ENTRY_SIZE)
        .ok_or(NipcError::Overflow)?;
    size = size
        .checked_add(7)
        .map(|v| v & !7)
        .ok_or(NipcError::Overflow)?;
    let item_size = CGROUPS_LOOKUP_ITEM_HDR_SIZE
        .checked_add(path.len())
        .and_then(|v| v.checked_add(2))
        .ok_or(NipcError::Overflow)?;
    size = size.checked_add(item_size).ok_or(NipcError::Overflow)?;

    let mut buf = vec![0u8; size];
    let mut builder = CgroupsLookupBuilder::new(&mut buf, 1, 0);
    builder.add(CGROUP_LOOKUP_OVERSIZED_ITEM, 0, path, b"", &[])?;
    let n = builder.finish()?;
    let view = CgroupsLookupResponseView::decode(&buf[..n])?;
    Ok(view.raw_item(0)?.to_vec())
}

pub type CgroupsLookupHandler = Arc<
    dyn for<'a> Fn(&CgroupsLookupRequestView, &mut CgroupsLookupBuilder<'a>) -> bool + Send + Sync,
>;

impl RawClient {
    /// Create a new client context bound to the cgroups-lookup service kind.
    /// Does NOT connect. Does NOT require the server to be running.
    pub fn new_cgroups_lookup(run_dir: &str, service_name: &str, config: ClientConfig) -> Self {
        Self::new_bound(run_dir, service_name, METHOD_CGROUPS_LOOKUP, config)
    }

    /// Blocking typed call: CGROUPS_LOOKUP method.
    ///
    /// The returned view is valid until the next typed call on this client.
    pub fn call_cgroups_lookup(
        &mut self,
        paths: &[&[u8]],
    ) -> Result<CgroupsLookupResponseView<'_>, NipcError> {
        self.call_cgroups_lookup_with_timeout(paths, 0)
    }

    /// Blocking typed call with an explicit timeout in milliseconds.
    ///
    /// A zero timeout uses the client's context-level default.
    pub fn call_cgroups_lookup_with_timeout(
        &mut self,
        paths: &[&[u8]],
        timeout_ms: u32,
    ) -> Result<CgroupsLookupResponseView<'_>, NipcError> {
        self.validate_method(METHOD_CGROUPS_LOOKUP)?;
        if paths.len() > self.max_logical_lookup_items as usize {
            return Err(NipcError::Overflow);
        }
        self.ensure_ready_for_logical_lookup()?;

        let mut raw_items: Vec<Vec<u8>> = Vec::with_capacity(paths.len());
        let mut start = 0usize;
        let mut generation: Option<u64> = None;
        let mut subcalls = 0u32;
        let deadline = self.lookup_deadline(timeout_ms)?;
        loop {
            let (req_count, req_size) = match cgroups_lookup_next_request_count(
                &paths[start..],
                self.session_max_request_payload_bytes(),
            ) {
                Ok(next) => next,
                Err(err) if err == NipcError::Overflow && start < paths.len() => {
                    let one_item_size = cgroups_lookup_request_size(&paths[start..start + 1])?;
                    if self.ensure_lookup_request_capacity(one_item_size)? {
                        continue;
                    }
                    raw_items.push(cgroups_lookup_oversized_request_item(paths[start])?);
                    start += 1;
                    if start < paths.len() {
                        continue;
                    }
                    cgroups_lookup_next_request_count(
                        &paths[start..],
                        self.session_max_request_payload_bytes(),
                    )?
                }
                Err(err) => return Err(err),
            };
            let req_paths = &paths[start..start + req_count];

            let req_len = {
                let req_buf = self.request_scratch(req_size);
                protocol::encode_cgroups_lookup_request(req_paths, req_buf)?
            };
            let remaining_timeout = self.lookup_remaining_timeout(deadline)?;
            subcalls += 1;
            if subcalls > self.max_logical_lookup_subcalls {
                return Err(NipcError::Overflow);
            }
            let response = self.raw_call_with_retry_timeout(
                METHOD_CGROUPS_LOOKUP,
                req_len,
                RawCallKind::single(),
                remaining_timeout,
            )?;
            let final_size = {
                let view = CgroupsLookupResponseView::decode(self.response_payload(response)?)?;
                if let Some(expected_generation) = generation {
                    if view.generation != expected_generation {
                        return Err(NipcError::BadLayout);
                    }
                } else {
                    generation = Some(view.generation);
                }
                if view.item_count != req_paths.len() as u32 {
                    return Err(NipcError::BadItemCount);
                }
                let mut payload_exceeded_at: Option<usize> = None;
                for (i, expected) in req_paths.iter().enumerate() {
                    let item = view.item(i as u32)?;
                    if item.path.as_bytes() != *expected {
                        return Err(NipcError::BadLayout);
                    }
                    if item.status == CGROUP_LOOKUP_PAYLOAD_EXCEEDED {
                        payload_exceeded_at = Some(i);
                        break;
                    }
                    raw_items.push(view.raw_item(i as u32)?.to_vec());
                }
                if let Some(first) = payload_exceeded_at {
                    for (i, expected) in req_paths.iter().enumerate().skip(first) {
                        let item = view.item(i as u32)?;
                        if item.path.as_bytes() != *expected
                            || item.status != CGROUP_LOOKUP_PAYLOAD_EXCEEDED
                        {
                            return Err(NipcError::BadLayout);
                        }
                    }
                    if first == 0 {
                        return Err(NipcError::Overflow);
                    }
                    start += first;
                    continue;
                }
                start += req_count;
                if start < paths.len() {
                    continue;
                }
                if raw_items.len() != paths.len() {
                    return Err(NipcError::BadItemCount);
                }
                let size = lookup_raw_response_size(
                    CGROUPS_LOOKUP_RESP_HDR_SIZE,
                    raw_items.len(),
                    &raw_items,
                )?;
                if size > self.max_logical_lookup_response_bytes as usize {
                    return Err(NipcError::Overflow);
                }
                size
            };
            self.transport_buf.resize(final_size, 0);
            let refs: Vec<&[u8]> = raw_items.iter().map(Vec::as_slice).collect();
            let n = protocol::encode_cgroups_lookup_raw_response(
                &refs,
                generation.unwrap_or(0),
                &mut self.transport_buf[..final_size],
            )?;
            return CgroupsLookupResponseView::decode(&self.transport_buf[..n]);
        }
    }
}

pub fn cgroups_lookup_dispatch(handler: CgroupsLookupHandler) -> DispatchHandler {
    Arc::new(move |request, response_buf| {
        protocol::dispatch_cgroups_lookup(request, response_buf, |request, builder| {
            handler(request, builder)
        })
        .map_err(dispatch_error_from_protocol)
    })
}
